optimizely/helpers/condition.py

```diff
@@ -12,115 +12,348 @@
+      None: if conditions couldn't be evaluated
+
+    """
+
+    if isinstance(conditions, list):
+      if conditions[0] in self.DEFAULT_OPERATOR_TYPES:
+        return self.EVALUATORS_BY_OPERATOR_TYPE[conditions[0]](self, conditions[1:], leaf_evaluator)
+      else:
+        # assume OR when operator is not explicit
+        return self.EVALUATORS_BY_OPERATOR_TYPE[ConditionOperatorTypes.OR](self, conditions, leaf_evaluator)
+
+    leaf_condition = conditions
+    return leaf_evaluator(leaf_condition)
+
+
+class CustomAttributeConditionEvaluator(object):
+  """ Class encapsulating methods to be used in audience leaf condition evaluation. """
+
+  CUSTOM_ATTRIBUTE_CONDITION_TYPE = 'custom_attribute'
+
+  MATCH_TYPES = [
+    ConditionMatchTypes.EXACT,
+    ConditionMatchTypes.EXISTS,
+    ConditionMatchTypes.GREATER_THAN,
+    ConditionMatchTypes.LESS_THAN,
+    ConditionMatchTypes.SUBSTRING
+  ]
+
+  def __init__(self, condition_data, attributes):
+    self.condition_data = condition_data
+    self.attributes = attributes or {}
+
+  def is_finite(self, value):
+    """ Method to validate if the given value is a number and not one of NAN, INF, -INF
+
+    Args:
+      value: Value to be validated
+
+    Returns:
+      Boolean: True if value is a number and not NAN, INF or -INF else False
+    """
+    if not isinstance(value, (numbers.Integral, float)):
+      # numbers.Integral instead of int to accomodate long integer in python 2
+      return False
+
+    if isinstance(value, bool):
+      # bool is a subclass of int
       return False
 
-    return not self.evaluate(single_condition[0])
+    if math.isnan(value) or math.isinf(value):
+      return False
+
+    return True
+
+  def is_value_valid_for_exact_conditions(self, value):
+    """ Method to validate if the value is valid for exact match type evaluation
+
+    Args:
+      value: Value to validate
+
+    Returns:
+      Boolean: True if value is a string type, or a boolean, or is finite. Otherwise False
+    """
+    if isinstance(value, string_types) or isinstance(value, bool) or self.is_finite(value):
+      return True
+
+    return False
+
+  def are_values_same_type(self, first_val, second_val):
+    """ Method to verify that both values belong to same type. Float and integer are
+    considered as same type.
+
+    Args:
+      first_val: Value to validate
+      second_Val: Value to validate
+
+    Returns:
+      Boolean: True if both values belong to same type. Otherwise False
+    """
+
+    first_val_type = type(first_val)
+    second_val_type = type(second_val)
+
+    # use isinstance to accomodate Python 2 unicode and str types
+    if isinstance(first_val, string_types) and isinstance(second_val, string_types):
+      return True
+
+    # Compare types if one of the values is bool because bool is a subclass on Integer
+    if isinstance(first_val, bool) or isinstance(second_val, bool):
+      return first_val_type == second_val_type
```

Comment:
In that case if only 1 of values is a bool type and the other is float or int
it will return true in the following if block as bool is a subclass of int

Reply:
Oh wow, you're right. So tricky! I haven't looked at the unit tests yet but hopefully we have test cases for this. The compatibility suite can't test the specifics of Python data types